### ISSUE TITLE
Images fail to download when using processor

### ISSUE THREAD

**Issue by zachwaugh** [pinterest/PINRemoteImage#372]
I noticed a regression on master and tracked it down to this commit https://github.com/pinterest/PINRemoteImage/commit/3781e4bc39466412ac745770eb4533fe90e7bf77. When downloading a bunch of images using `processorKey` and `processor`, I don't receive any completion callbacks for the majority of them. The first few requests work, but the rest don't. For the majority that don't work, it appears the network request is never made looking at the server logs. I'm still digging into to try to understand all the changes, but thought I would open the issue in case something jumps out.

**Comment by garrettmoon** [comment]
@zachwaugh if you're able to reproduce, can you try again with master? I fixed a deadlock #366.

Separately, if you can upload a sample project I'd love to debug!

**Comment by zachwaugh** [comment]
@garrettmoon I'm still working on a sample project. I made one, but so far it's not reproducing the bug. I built against master and added some more logging in my own project, and I think there is a deadlock related to canceling a download. I believe it's here - https://github.com/pinterest/PINRemoteImage/blob/master/Source/Classes/PINRemoteImageProcessorTask.m#L19-L20, we're calling `self.downloadTaskUUID` within a lock and that method also locks. I'm not sure why it doesn't always happen, but seems we can do direct ivar access there
